diff -Naur postgresql-10.1.orig/src/tools/msvc/Mkvcbuild.pm postgresql-10.1/src/tools/msvc/Mkvcbuild.pm
--- postgresql-10.1.orig/src/tools/msvc/Mkvcbuild.pm	2017-12-06 19:40:59.694628794 +0300
+++ postgresql-10.1/src/tools/msvc/Mkvcbuild.pm	2017-12-06 20:27:23.879553458 +0300
@@ -615,9 +615,9 @@
 			}
 		}
 		$plperl->AddReference($postgres);
-		my $perl_path = $solution->{options}->{perl} . '\lib\CORE\perl*.lib';
+		my $perl_path = $solution->{options}->{perl} . '\lib\CORE\*perl*.*';
 		my @perl_libs =
-		  grep { /perl\d+.lib$/ } glob($perl_path);
+		  grep { /perl\d+.(lib|a)$/ } glob($perl_path);
 		if (@perl_libs == 1)
 		{
 			$plperl->AddLibrary($perl_libs[0]);
