diff --git a/src/include/port.h b/src/include/port.h
index 88603c4def9..3530aa386df 100644
--- a/src/include/port.h
+++ b/src/include/port.h
@@ -17,6 +17,41 @@
 #include <netdb.h>
 #include <pwd.h>
 
+#if _MSC_VER > 1800
+	//From VS2012.
+	typedef struct localerefcount
+	{
+		char *locale;
+		wchar_t *wlocale;
+		int *refcount;
+		int *wrefcount;
+	} locrefcount;
+
+	//From VS2012.
+	typedef struct __crt_locale_data
+	{
+		int refcount;
+		unsigned int lc_codepage;
+		unsigned int lc_collate_cp;
+		unsigned int lc_time_cp;
+		locrefcount lc_category[6];
+		int lc_clike;
+		int mb_cur_max;
+		int * lconv_intl_refcount;
+		int * lconv_num_refcount;
+		int * lconv_mon_refcount;
+		struct lconv * lconv;
+		int * ctype1_refcount;
+		unsigned short * ctype1;
+		const unsigned short * pctype;
+		const unsigned char * pclmap;
+		const unsigned char * pcumap;
+		struct __lc_time_data * lc_time_curr;
+		wchar_t * locale_name[6];
+	} threadlocinfo;
+#endif
+
+
 /*
  * Windows has enough specialized port stuff that we push most of it off
  * into another file.
