diff --git a/src/tools/msvc/MSBuildProject.pm b/src/tools/msvc/MSBuildProject.pm
index d7638b4..ccf1f76 100644
--- a/src/tools/msvc/MSBuildProject.pm
+++ b/src/tools/msvc/MSBuildProject.pm
@@ -345,6 +345,9 @@ EOF
 		$d =~ s/__CFGNAME__/$cfgname/g;
 		print $f "      <ModuleDefinitionFile>$d</ModuleDefinitionFile>\n";
 	}
+	if ($self->{name} =~ /plperl/ and $self->{platform} eq 'Win32') {
+		print $f "      <ImageHasSafeExceptionHandlers>false</ImageHasSafeExceptionHandlers>\n";
+	}
 	print $f <<EOF;
     </Link>
     <ResourceCompile>
diff --git a/src/tools/msvc/Mkvcbuild.pm b/src/tools/msvc/Mkvcbuild.pm
index 21faa58..7012bfa 100644
--- a/src/tools/msvc/Mkvcbuild.pm
+++ b/src/tools/msvc/Mkvcbuild.pm
@@ -572,8 +578,8 @@ sub mkvcbuild
 		}
 		$plperl->AddReference($postgres);
 		my @perl_libs =
-		  grep { /perl\d+.lib$/ }
-		  glob($solution->{options}->{perl} . '\lib\CORE\perl*.lib');
+		  grep { /perl\d+\.(lib|a)$/ }
+		  glob($solution->{options}->{perl} . '\lib\CORE\*.*');
 		if (@perl_libs == 1)
 		{
 			$plperl->AddLibrary($perl_libs[0]);
